Dereference after null reference (CID:1124543)

nishith-vihar opened the pull request:
'this' pointer was being dereferenced after null check. This change avoids it.

Change-Id: I7dedee44c08df481d2a037eb601f3d5c4d9284f5
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

nishith-vihar committed:
Dereference after null reference (CID:1124543)

'this' pointer was being dereferenced after null check. This change avoids it.

Change-Id: I7dedee44c08df481d2a037eb601f3d5c4d9284f5
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

schaffung commented:
/run regression

nik-redhat commented on xlators/mgmt/glusterd/src/glusterd-handshake.c:
```diff
@@ -1948,7 +1948,7 @@ gd_validate_peer_op_version(xlator_t *this, glusterd_peerinfo_t *peerinfo,
         gf_smsg("glusterd", GF_LOG_ERROR, errno, GD_MSG_XLATOR_NOT_DEFINED,
                 NULL);
-        goto out;
+        return ret;
```
Simply using `GF_ASSERT(this)` at line:1941 would be a better fix, I think.

srijan-sivakumar replied:
+1

xhernandez replied:
I don't think so. Our `GF_ASSERT()` implementation for production builds is simply an `if` and a log message if the condition is false, but the execution flow continues. From the point of view of coverity, this has the same problem: we would have checked for NULL and then dereferenced it.

Either we assume that `this` is never NULL (as it should be) and we don't check it, or we correctly handle the case. We could also use `GF_ABORT()` when it's NULL, which causes a core dump and stops the program, but this will already happen when it's first dereferenced.

schaffung replied:
Actually this function is invoked by __glusterd_mgmt_hndsk_version_cbk wherein this is derived from THIS. So I think GF_ASSERT isn't required.

mykaul replied:
...and it would have crashed here - https://github.com/gluster/glusterfs/blob/3813ffafb12d20d2b728d43b3f0eb5201e6fb7c2/xlators/mgmt/glusterd/src/glusterd-handshake.c#L2125 otherwise... Surely can be removed.

(but I'd also make gd_validate_peer_op_version() a static function).
(and of course, so should the whole chain of functions in this flow, but this should be in a different patch...)

nik-redhat replied:
At other places too, we are using `GF_ASSERT` to check for NULL and it does crash if it's NULL.
Though, that's a different aspect that we might never get `this` as NULL here.

Edit: But, yes the execution continues after that.

nishith-vihar committed:
Dereference after null reference (CID:1124543)
'this' pointer was being dereferenced after null check. 'if' branch with '!this' as condition is removed as it is redundant in the flow.

Change-Id: I4804a894d0d69fc8d75affbfce45255876d2dd4a
Updates: #1060
Signed-off-by: Nishith Vihar Sakinala <nsakinal@redhat.com>

schaffung commented:
/run regression

schaffung commented:
/recheck smoke

schaffung commented:
/run regression
